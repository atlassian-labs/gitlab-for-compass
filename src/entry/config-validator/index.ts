import { ConfigValidatorResult, ConfigValidatorResponse } from '@atlassian/forge-graphql';

import { getConnectedGroups } from '../../services/group';

export const configValidator = async (): Promise<ConfigValidatorResult> => {
  const connectedGroups = await getConnectedGroups();
  const appConfigured = connectedGroups && connectedGroups.length > 0;

  const response = new ConfigValidatorResponse(appConfigured);
  return response.build();
};
