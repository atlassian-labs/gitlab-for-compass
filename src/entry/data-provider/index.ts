import {
  BuiltinMetricDefinitions,
  DataProviderEventTypes,
  DataProviderResponse,
  DataProviderResult,
} from '@atlassian/forge-graphql';

import { DataProviderPayload } from './types';
import { getProjectDataFromUrl } from '../../services/data-provider-link-parser';
import { getTrackingBranchName } from '../../services/get-tracking-branch';
import { getBackfillData } from '../../services/get-backfill-data';
import { parse } from '../../utils/parse-ari';

export const dataProvider = async (request: DataProviderPayload): Promise<DataProviderResult> => {
  try {
    parse(request.ctx.cloudId);
  } catch {
    console.error('Invalid cloudId.');
    return null;
  }

  const {
    project: { id: projectId, default_branch: defaultBranch, name: projectName },
    groupToken,
  } = await getProjectDataFromUrl(request.url);

  if (!projectId) {
    console.warn('Cannot get GitLab project data by provided link.');
    return null;
  }

  const trackingBranch = await getTrackingBranchName(groupToken, projectId, defaultBranch);

  const {
    builds,
    deployments,
    metrics: { mrCycleTime, openMergeRequestsCount },
  } = await getBackfillData(groupToken, projectId, projectName, trackingBranch);

  const response = new DataProviderResponse(projectId.toString(), {
    eventTypes: [DataProviderEventTypes.BUILDS, DataProviderEventTypes.DEPLOYMENTS],
    builtInMetricDefinitions: [
      {
        name: BuiltinMetricDefinitions.WEEKLY_DEPLOYMENT_FREQUENCY_28D,
        derived: true,
      },
      {
        name: BuiltinMetricDefinitions.PULL_REQUEST_CYCLE_TIME_AVG_LAST_10,
        derived: false,
      },
      {
        name: BuiltinMetricDefinitions.BUILD_TIME_AVG_LAST_10,
        derived: true,
      },
      {
        name: BuiltinMetricDefinitions.OPEN_PULL_REQUESTS,
        derived: false,
      },
      {
        name: BuiltinMetricDefinitions.BUILD_SUCCESS_RATE,
        derived: true,
      },
    ],
    customMetricDefinitions: [],
  });

  return response
    .addBuilds(builds)
    .addDeployments(deployments)
    .addBuiltInMetricValue(BuiltinMetricDefinitions.PULL_REQUEST_CYCLE_TIME_AVG_LAST_10, mrCycleTime)
    .addBuiltInMetricValue(BuiltinMetricDefinitions.OPEN_PULL_REQUESTS, openMergeRequestsCount)
    .build();
};
