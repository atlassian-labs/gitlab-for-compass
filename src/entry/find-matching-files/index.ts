import { ProjectFile } from 'src/types';
import { FindMatchingFilesPayload, FindMatchingFilesResponse } from './types';
import { getProjectDataFromUrl } from '../../services/data-provider-link-parser';
import { listFiles } from '../../client/gitlab';
import { isPackageDependenciesM3Enabled } from '../../services/feature-flags';

const MAX_NUM_FILES_PER_PAGE = 100;

const getFiles = async (groupToken: string, projectId: number, path?: string) => {
  let shouldContinue = true;
  let pageToken;
  const topLevelFiles: ProjectFile[] = [];

  while (shouldContinue) {
    const files: ProjectFile[] = await listFiles(
      groupToken,
      projectId,
      path,
      undefined,
      pageToken,
      MAX_NUM_FILES_PER_PAGE,
      false,
    );

    topLevelFiles.push(...files);

    if (files.length < MAX_NUM_FILES_PER_PAGE) {
      // there are no more files to paginate
      shouldContinue = false;
    } else {
      pageToken = files[files.length - 1].id;
    }
  }

  return topLevelFiles;
};

const getTreeProjectFiles = (files: ProjectFile[]) => {
  return files.filter((file) => file.type === 'tree');
};

const getFileBlobs = (files: ProjectFile[]) => {
  return files.filter((file) => file.type === 'blob');
};

// Searches 2 levels deep for files
export const findMatchingFiles = async (payload: FindMatchingFilesPayload): Promise<FindMatchingFilesResponse> => {
  if (!isPackageDependenciesM3Enabled()) {
    return {
      success: false,
      errorMessage: 'Feature not implemented.',
      statusCode: 405,
      files: [],
    };
  }
  try {
    if (payload.fileName.containsOneOf.length === 0 && payload.fileName.equalsOneOf.length === 0) {
      return {
        success: false,
        errorMessage: 'No file names provided.',
        statusCode: 400,
        files: [],
      };
    }

    const projectData = await getProjectDataFromUrl(payload.repoUrl);
    if (projectData == null) {
      return {
        success: false,
        errorMessage: 'Error parsing repository URL.',
        statusCode: 400,
        files: [],
      };
    }

    const { project, groupToken } = projectData;

    let allFiles: ProjectFile[] = [];
    const topLevelProjectFiles = await getFiles(groupToken, project.id);
    const topLevelProjectFileBlobs = getFileBlobs(topLevelProjectFiles);
    const treeProjectFiles = getTreeProjectFiles(topLevelProjectFiles);

    allFiles = [...topLevelProjectFileBlobs];

    for (const file of treeProjectFiles) {
      const files = await getFiles(groupToken, project.id, file.path);
      const fileBlobs = getFileBlobs(files);
      allFiles = [...allFiles, ...fileBlobs];
    }

    const matchingFiles = allFiles.filter(
      (file) => payload.fileName.containsOneOf.includes(file.name) || payload.fileName.equalsOneOf.includes(file.name),
    );

    console.log('matchingFiles', matchingFiles);
    return {
      success: true,
      files: matchingFiles.map((file) => ({
        path: file.path,
        metadata: {},
      })),
      statusCode: 200,
    };
  } catch (e) {
    console.log('Failed to find matching files', e.message);
    return {
      success: false,
      errorMessage: e.message,
      statusCode: e.cause ?? 500,
      files: [],
    };
  }
};
