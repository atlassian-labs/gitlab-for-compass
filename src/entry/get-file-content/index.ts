import { storage } from '@forge/api';
import { GetFileContentsPayload, GetFileContentsResponse } from './types';
import { getRawFileContent } from '../../client/gitlab';
import { STORAGE_SECRETS } from '../../constants';
import { GitlabHttpMethodError } from '../../models/errors';

export const getFileContent = async (payload: GetFileContentsPayload): Promise<GetFileContentsResponse> => {
  try {
    const groupToken = await storage.getSecret(`${STORAGE_SECRETS.GROUP_TOKEN_KEY_PREFIX}${payload.groupId}`);
    const contents = await getRawFileContent(groupToken, payload.projectId, payload.filePath, payload.ref);

    return {
      success: true,
      file: {
        contents,
      },
      statusCode: 200,
    };
  } catch (error) {
    if (error instanceof GitlabHttpMethodError) {
      return {
        success: false,
        errorMessage: error.statusText,
        statusCode: error.status,
      };
    }
    console.error({ message: 'Error performing getFileContents' });
    return {
      success: false,
      errorMessage: error.message,
      statusCode: 500,
    };
  }
};
