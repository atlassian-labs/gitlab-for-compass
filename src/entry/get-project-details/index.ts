import { GetProjectDetailsPayload, GetProjectDetailsResponse } from './types';
import { getProjectDataFromUrl } from '../../services/data-provider-link-parser';
import { getProjectTrackingBranch } from '../../services/get-tracking-branch';
import { GitlabAPIProject } from '../../types';

export const getProjectDetails = async (payload: GetProjectDetailsPayload): Promise<GetProjectDetailsResponse> => {
  try {
    let projectDataResult;
    try {
      projectDataResult = await getProjectDataFromUrl(payload.projectUrl);
    } catch (e) {
      console.warn('Cannot get GitLab project data from provided url.');
      return {
        success: false,
        errorMessage: 'project not found',
        statusCode: 404,
      };
    }

    const {
      project: { id: projectId, default_branch: defaultBranch },
      groupToken,
      groupId,
    } = projectDataResult;

    const branchResult = await getProjectTrackingBranch(groupToken, projectId, defaultBranch);
    if (!branchResult) {
      console.warn('Cannot get GitLab project branch data. Branch not found.');
      return {
        success: false,
        errorMessage: 'branch not found',
        statusCode: 404,
      };
    }
    const {
      name,
      commit: { id: commitId },
    } = branchResult;

    return {
      success: true,
      project: {
        projectId,
        groupId,
        defaultBranch: name,
        shaOnDefaultBranch: commitId,
      },
      statusCode: 200,
    };
  } catch (error) {
    console.error({ message: 'Error performing getProjectDetails' });
    return {
      success: false,
      errorMessage: error.message,
      statusCode: 500,
    };
  }
};
