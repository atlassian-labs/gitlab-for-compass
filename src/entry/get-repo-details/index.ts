import { GetRepoDetailsPayload, GetRepoDetailsResponse } from './types';
import { getProjectDataFromUrl } from '../../services/data-provider-link-parser';
import { getProjectTrackingBranch } from '../../services/get-tracking-branch';

export const getRepoDetails = async (payload: GetRepoDetailsPayload): Promise<GetRepoDetailsResponse> => {
  try {
    const projectDataResult = await getProjectDataFromUrl(payload.projectUrl);
    if (!projectDataResult) {
      console.warn('GitLab project not found.');
      return {
        success: false,
        errorMessage: 'project not found',
        statusCode: 404,
      };
    }

    const {
      project: { id: projectId, default_branch: defaultBranch },
      groupToken,
      groupId,
    } = projectDataResult;

    const branchResult = await getProjectTrackingBranch(groupToken, projectId, defaultBranch);
    if (!branchResult) {
      console.warn('Cannot get GitLab project branch data. Branch not found.');
      return {
        success: false,
        errorMessage: 'branch not found',
        statusCode: 404,
      };
    }
    const {
      name,
      commit: { id: commitId },
    } = branchResult;

    return {
      success: true,
      project: {
        projectId,
        groupId,
        defaultBranch: name,
        shaOnDefaultBranch: commitId,
      },
      statusCode: 200,
    };
  } catch (error) {
    console.error({ message: 'Error performing getProjectDetails' });
    return {
      success: false,
      errorMessage: error.message,
      statusCode: 500,
    };
  }
};
