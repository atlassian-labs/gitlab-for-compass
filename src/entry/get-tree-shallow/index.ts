import { storage } from '@forge/api';
import { GetTreeShallowPayload, GetTreeShallowResponse } from './types';
import { listFiles } from '../../client/gitlab';
import { STORAGE_SECRETS } from '../../constants';

export const getTreeShallow = async (payload: GetTreeShallowPayload): Promise<GetTreeShallowResponse> => {
  try {
    const groupToken = await storage.getSecret(`${STORAGE_SECRETS.GROUP_TOKEN_KEY_PREFIX}${payload.groupId}`);
    const projectFiles = await listFiles(
      groupToken,
      payload.projectId,
      payload.path,
      payload.ref,
      payload.pageToken,
      payload.pageSize,
      payload.recursive,
    );

    return {
      success: true,
      files: projectFiles,
      statusCode: 200,
    };
  } catch (error) {
    console.error({ message: 'Error performing listFilesInPath' });
    return {
      success: false,
      errorMessage: error.message,
      statusCode: 500,
    };
  }
};
