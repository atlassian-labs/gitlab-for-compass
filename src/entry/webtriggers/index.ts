export { processGitlabEvent } from './process-gitlab-event';
