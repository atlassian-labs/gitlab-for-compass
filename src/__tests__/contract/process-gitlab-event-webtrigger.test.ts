/* eslint-disable import/order */
import { storage, mockForgeApi } from '../helpers/forge-helper';
/* eslint-disable import/first */
mockForgeApi();

import { processGitlabEvent } from '../../entry/webtriggers';
import { pipelineWebhookFixture } from '../fixtures/build-webhook-payload';

const MOCK_TOKEN_SECRET = 'kokokokokokokokok';

describe('Process gitlab webtrigger', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('successfully parse webhook event', async () => {
    storage.get.mockResolvedValue(MOCK_TOKEN_SECRET);
    const resp = await processGitlabEvent(
      {
        body: JSON.stringify({ body: 'some data' }),
        headers: { 'x-gitlab-token': [MOCK_TOKEN_SECRET] },
        queryParameters: { groupId: [12345] },
      },
      { installContext: 'ari:cloud:compass::site/00000000-0000-0000-0000-000000000000', principal: undefined },
    );

    expect(resp).toMatchSnapshot();
  });

  test('successfully parse webhook pipeline event', async () => {
    storage.get.mockResolvedValue(MOCK_TOKEN_SECRET);
    const resp = await processGitlabEvent(
      {
        body: JSON.stringify({ body: pipelineWebhookFixture }),
        headers: { 'x-gitlab-token': [MOCK_TOKEN_SECRET] },
        queryParameters: { groupId: [12345] },
      },
      { installContext: 'ari:cloud:compass::site/00000000-0000-0000-0000-000000000000', principal: undefined },
    );

    expect(resp).toMatchSnapshot();
  });
});
