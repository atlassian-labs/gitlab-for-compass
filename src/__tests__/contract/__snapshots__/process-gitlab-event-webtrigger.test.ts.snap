// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process gitlab webtrigger successfully parse webhook event 1`] = `
Object {
  "body": "{\\"message\\":\\"Processed webhook event\\",\\"success\\":true}",
  "headers": Object {
    "Content-Type": Array [
      "application/json",
    ],
  },
  "statusCode": 200,
}
`;

exports[`Process gitlab webtrigger successfully parse webhook pipeline event 1`] = `
Object {
  "body": "{\\"message\\":\\"Processed webhook event\\",\\"success\\":true}",
  "headers": Object {
    "Content-Type": Array [
      "application/json",
    ],
  },
  "statusCode": 200,
}
`;
